components/calendar/src/islamic.rs

```diff
@@ -308,6 +305,267 @@ impl DateTime<IslamicObservational> {
+    pub fn new() -> Self {
+        Self
+    }
+    // Lisp code reference: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L6957
+    fn saudi_criterion(date: RataDie) -> Option<bool> {
+        let sunset = Astronomical::sunset((date - 1).as_moment(), MECCA)?;
+        let tee = Location::universal_from_standard(sunset, MECCA);
+        let phase = Astronomical::lunar_phase(tee);
+        let moonlag = Astronomical::moonlag((date - 1).as_moment(), MECCA)?;
+
+        Some(phase > 0.0 && phase < 90.0 && moonlag > 0.0)
+    }
+    pub(crate) fn adjusted_saudi_criterion(date: RataDie) -> bool {
+        if let Some(x) = Self::saudi_criterion(date) {
+            x
+        } else {
+            false
+        }
+    }
+
+    fn saudi_new_month_on_or_before(date: RataDie) -> RataDie {
+        let moon =
+            libm::floor((Astronomical::lunar_phase_at_or_before(0.0, date.as_moment())).inner());
+        let age = date.to_f64_date() - moon;
+        let tau = if age <= 3.0 && !Self::adjusted_saudi_criterion(date) {
+            moon - 30.0
+        } else {
+            moon
+        };
+
+        next(RataDie::new(tau as i64), Self::adjusted_saudi_criterion)
+    }
+
+    // Lisp code reference: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L6996
+    #[allow(clippy::unwrap_used)]
+    fn saudi_islamic_from_fixed(date: RataDie) -> Date<UmmalQura> {
+        let crescent = Self::saudi_new_month_on_or_before(date);
+        let elapsed_months =
+            libm::round((crescent - FIXED_ISLAMIC_EPOCH_FRIDAY) as f64 / MEAN_SYNODIC_MONTH);
+        let year = (1.0 + div_rem_euclid_f64(elapsed_months, 12.0).0) as i32;
+        let month = (1.0 + div_rem_euclid_f64(elapsed_months, 12.0).1) as u8;
+        let day = ((date - crescent) + 1) as u8;
+
+        Date::try_new_ummalqura_date(year, month, day).unwrap()
+    }
+
+    // "Fixed" is a day count representation of calendars staring from Jan 1st of year 1 of the Georgian Calendar.
+    // The fixed date algorithms are from
+    // Dershowitz, Nachum, and Edward M. Reingold. _Calendrical calculations_. Cambridge University Press, 2008.
+    //
+    // Lisp code reference: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L6981
+    fn fixed_from_saudi_islamic(date: UmmalQuraDateInner) -> RataDie {
+        let year = date.0.year;
+        let month = date.0.month;
+        let day = date.0.day;
+
+        let midmonth = RataDie::new(libm::floor(
+            FIXED_ISLAMIC_EPOCH_FRIDAY.to_f64_date()
+                + (((year as f64 - 1.0) * 12.0 + month as f64 - 0.5) * MEAN_SYNODIC_MONTH),
+        ) as i64);
+        let first_day_of_month = Self::saudi_new_month_on_or_before(midmonth).to_i64_date();
+
+        RataDie::new(first_day_of_month + day as i64 - 1)
+    }
+
+    fn year_as_islamic(year: i32) -> types::FormattableYear {
+        types::FormattableYear {
+            era: types::Era(tinystr!(16, "ah")),
```

Comment:
This should return "islamic-umalqura", but when constructing both "islamic" and "ah" should be accepted

See https://tc39.es/proposal-intl-era-monthcode/#table-eras

also fix this for the observational calendar